Make a seperate tab for comments in wikis #1438

aspriya opened the pull request:
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

ananyo2012 commented:
@aspriya Can you post some screenshots?

aspriya commented:
Sure, Following image is a wiki show page included with a seperate tab icon for commenting.

![tab icon](https://user-images.githubusercontent.com/13878973/28240485-9df08f62-69a0-11e7-8ab6-049f9b8fd6e3.png)

-----
And following shows the inside of that tab. It contains previously entered comments and a form to enter comments.
![tab](https://user-images.githubusercontent.com/13878973/28240489-ab7ebed8-69a0-11e7-81b2-6e65b5afbf82.png)

ananyo2012 commented:
Wouldn't it be better if we keep the comments below the wiki content like in research notes and questions?

aspriya commented:
In my summer of code project adding a seperate tab for commenting is an one task. But In a future task I'm Bringing commenting functionality per subsection basis. You can view my task list from [here](https://github.com/publiclab/plots2/issues/1438). This is because @jywarren and I talked previously, before implementing inline commenting, it is good to implement commenting in a sepereta tab like this.
Thanx :smile:

aspriya commented:
This is failing because the last added commit has a test and it's failing. That's due to codes in previous commits are not yet merged.

jywarren commented:
What tests are failing? The changes listed don't seem like they'd cause trouble. Can you look in the Traivs tests (click Details) and copy in the exact error? I think you may have to either modify your own code or the tests to account for this. 

Looks good however!

jywarren commented:
Also can you list the # of comments next to the tab icon? Thanks!

ananyo2012 commented:
@aspriya Can you re add the test and check if it is passing?

aspriya commented:
I'm Hoping to do the tests in a seperate branch and send a PR. Is it ok?

ananyo2012 commented:
Well it will be good if you have tests for the feature along with code. Can
you try adding it once more so than we can inspect the problem? That will
give more confidence in merging the code.

On 17-Jul-2017 10:27 AM, "Ashan Priyadarshana" <notifications@github.com>
wrote:

I'm Hoping to do the tests in a seperate branch and send a PR. Is it ok?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/publiclab/plots2/pull/1539#issuecomment-315668278>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AKACR_iKtVqz8F8AYm8QWi1RdKuz9Ujoks5sOumzgaJpZM4OY796>
.

aspriya commented:
Thanks , Sure will do @ananyo2012

ananyo2012 commented:
Ok so I see the test failing [here](https://travis-ci.org/publiclab/plots2/builds/254695372#L2835)

ananyo2012 commented on test/functional/wiki_controller_test.rb:
```diff
@@ -572,4 +572,12 @@ def teardown
+  test "should render comment template when a comment=true parameter is passed in the get request" do
+    wiki = node(:wiki_page)
+    slug = wiki.path.gsub('/wiki/', '')
+    get "/wiki/data-logging?comments=true"
```
@aspriya The action corresponding to the url doesn't exist. Have a look at thw `wiki_page` fixture again. Also note that the wiki path and slug is different in this case

aspriya commented:
@ananyo2012 , thanks for your guidance, was stuck in, not running tests correctly due to some error in sunspot gem. Any way now the issue is solved and found why previously test was failing. The reason was the way I send the request and parameters is wrong in the test. It's corrected now and test is passing :smiley:

ananyo2012 commented on test/functional/wiki_controller_test.rb:
```diff
@@ -576,7 +576,8 @@ def teardown
     wiki = node(:wiki_page)
     slug = wiki.path.gsub('/wiki/', '')
-    get "/wiki/data-logging?comments=true"
+    puts slug
```
@aspriya Remove this line and we are good yo go!

aspriya replied:
Oh my bad, It's corrected

ananyo2012 commented:
Also remove [these lines](https://github.com/publiclab/plots2/blob/master/test/functional/comment_controller_test.rb#L56-L57) from tests since this is covered by this test now.

jywarren commented:
Thanks!!!

Ananyo - once this looks good to you, feel free to merge in!

jywarren commented on app/views/wiki/show.html.erb:
```diff
@@ -64,11 +64,15 @@
-      </div>
-      <% if params[:raw] == 'true' %>
-        <div style="display:none;" id="content-raw-markdown" class="pl-content wiki"><%= raw insert_extras(@revision.body_raw) %></div>
+      <% if params[:comments] == 'true' %>
```
maybe we should rework this for a new route, like `/wiki/_name_/comments`? instead of a parameter? You could add to `config/routes.rb` and direct it to a new action that's just for comments in `comment_controller.rb`. Thoughts?

aspriya replied:
Ah sure, but then what I will do is render the `show` action of `wiki_controller` from that newly created action in `comment_controller.rb` while passing the `id` of the wiki ?. 

And in the `show` action of `wiki_controller` I will check the controller and action it was rendered from and, if it was rendered from the action we just created within the `comment_controller`, I will render the comment tab ?

ananyo2012 replied:
@aspriya You got it right. Reuse the `wiki/show` template. If the `action` is `show` render the wiki body else the `notes/comments` template.

jywarren commented:
Yes, or you could make the tabs into a partial template and have a
wiki/comments.html.erb which also includes it. Either way!

On Jul 23, 2017 4:16 PM, "Ashan Priyadarshana" <notifications@github.com>
wrote:

*@aspriya* commented on this pull request.
------------------------------

In app/views/wiki/show.html.erb
<https://github.com/publiclab/plots2/pull/1539#discussion_r128920790>:

> @@ -64,11 +64,15 @@
     <% end %>

     <div class="tab-pane active" id="tab-overview">
-      <div <% unless @node.has_tag('style:wide')
%>style="max-width:800px;"<% end %> id="content" class="pl-content
wiki">
-        <%= raw auto_link(insert_extras(@revision.render_body),
sanitize: false) %>
-      </div>
-      <% if params[:raw] == 'true' %>
-        <div style="display:none;" id="content-raw-markdown"
class="pl-content wiki"><%= raw insert_extras(@revision.body_raw)
%></div>
+      <% if params[:comments] == 'true' %>

Ah sure, but then what I will do is render the show action of
wiki_controller from that newly created action in comment_controller.rb
while passing the id of the wiki ?.

And in the show action of wiki_controller I will check the controller and
action it was rendered from and, if it was rendered from the action we just
created within the comment_controller, I will render the comment tab ?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/publiclab/plots2/pull/1539#discussion_r128920790>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AABfJ1LfuleMBX1ZWmH2BcDj91gRL-isks5sQ1XIgaJpZM4OY796>
.

aspriya commented:
@jywarren, but if I create a different action within `comment_controller`, then I would have to rewrite all the codes within `wiki/show` action in this newly added action in `comment_controller` and at the end add a line like `render '/wiki/show'` to render the wiki/show template. Is it ok?  

Then as previously stated I can check the controller and action which the `wiki/show` view is rendering and by that can decide, to show or not to show the comment tab.

jywarren commented:
You can avoid repetition by calling the show action as a function from
within your new action just by using the command "show" -- but just be sure
your functional test confirms that both actions work as expected.

On Jul 23, 2017 5:40 PM, "Ashan Priyadarshana" <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren>, but if I create a different
> action within comment_controller, then I would have to rewrite all the
> codes within wiki/show action in this newly added action in
> comment_controller and at the end add a line like render '/wiki/show' to
> render the wiki/show template. Is it ok?
>
> Then as previously stated I can check the controller and action which the
> wiki/show view is rendering and by that can decide, to show or not to
> show the comment tab.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1539#issuecomment-317261828>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJz9mKcnbJxtpu9tO5jhElKpRNqG6ks5sQ2l0gaJpZM4OY796>
> .
>

aspriya commented:
ah yes, but the problem was I have written this within the comment controller. So I couldn't call the `show` action in `wiki_controller` from `comments_controller`. Any way now I have changed it to wiki controller and modified the code. changed the tests according to that also. Can you look at the PR again?

jywarren commented on config/routes.rb:
```diff
@@ -53,6 +53,8 @@
   match 'home' => 'home#front'
   resources :relationships, only: [:create, :destroy]
 
+  get '/wiki/:id/comments', to: 'wiki#wiki_comments'
```
This `wiki#wiki_comments` seems redundant - -any problem using just `wiki#comments` and action name just `comments`? sorry if I miss something here.

aspriya replied:
Oh my bad, this `wiki_comment` were used when the action was within `comments_controller`. @jywarren is right. As this is within `wiki_controller` just `comments` is enough and meaningfull. 
Thanks.

jywarren commented:
One request, otherwise looks good!

ananyo2012 commented:
Great! Merging!

jywarren commented:
Wow this will be fun to test out on the live site. In the next few days I
hope!

On Jul 26, 2017 10:55 AM, "Ananyo Maiti" <notifications@github.com> wrote:

Merged #1539 <https://github.com/publiclab/plots2/pull/1539>.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/publiclab/plots2/pull/1539#event-1180053246>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AABfJ0ddToR3PqIIn3t2yMaGUzNXoNUfks5sR1N4gaJpZM4OY796>
.

jywarren commented:
This works! It's live on the site now!

Great work!!! I also tested mentioning someones name with `@jywarren` and it worked -- got an email alert.

aspriya commented:
Wow. Great to here. Thanks @jywarren and @ananyo2012 for guidance. Can't wait to show this to friends at college :smiley:
